### PR TITLE
fix(mads): add mutex when checking if reconcile is needed and reconciling

### PR DESCRIPTION
**Pull request by lobkovilya**
This is a straightforward fix for the MADS problem we see in unit tests. But we probably don't need `NeedsReconciliation` at all since we fixed watchdogs to reconcile before the first tick https://github.com/kumahq/kuma/pull/11105

CC @Automaat @michaelbeaumont 

### Checklist prior to review

- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/mads/v1/reconcile/reconciler.go [hunk 1]
```diff
@@ -18,12 +18,14 @@ func NewReconciler(hasher envoy_cache.NodeHash, cache util_xds_v3.SnapshotCache,
...
 		knownClientIds:      map[string]bool{},
 		knownClientIdsMutex: &sync.Mutex{},
+		cacheMutex:          &sync.Mutex{},
 	}
 }
 
 type reconciler struct {
 	hasher              envoy_cache.NodeHash
 	cache               util_xds_v3.SnapshotCache
+	cacheMutex          *sync.Mutex
```
Comment (lobkovilya) [inline pkg/mads/v1/reconcile/reconciler.go:28]: ah probably should rename the mutex to something else

**Comment by Automaat** [comment]
> But we probably don't need NeedsReconciliation at all 
 
I am not sure about this, MADS works in a weird way since it is responding to http requests not relying entirely on watchdogs.

**Review by Automaat — APPROVED** [review]

**Review by lahabana — APPROVED** [review]
This is simple fix. The more complex fix is trickier because of these reconcilers running concurrently makes 0 sense. So this will work for now

**Review by slonka — APPROVED** [review]

**Comment by Icarus9913** [comment]
> I am not sure about this, MADS works in a weird way since it is responding to http requests not relying entirely on watchdogs.

Because the watchdog runs in goroutine and we need every interval to generate the XDS. And each http request gets response from go-control-plane `fetch` method.

Although we got that first tick reconcile from #11105, but you can imagine that the request happens between the first and second tick to fetch xds configurations then it would not found something new.

**Pull request merged**